multimixer-128/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+                ]);
+                k[i] = u32::from_ne_bytes([
+                    self.key_blocks[self.block_index][i * 4 + 16],
+                    self.key_blocks[self.block_index][i * 4 + 17],
+                    self.key_blocks[self.block_index][i * 4 + 18],
+                    self.key_blocks[self.block_index][i * 4 + 19],
+                ]);
+            }
+            (h, k)
+        };
+
+        for i in 0..4 {
+            x[i] = u32::from_ne_bytes([
+                message_block[i * 4],
+                message_block[1 + i * 4],
+                message_block[2 + i * 4],
+                message_block[3 + i * 4],
+            ]);
+            y[i] = u32::from_ne_bytes([
+                message_block[16 + i * 4],
+                message_block[17 + i * 4],
+                message_block[18 + i * 4],
+                message_block[19 + i * 4],
+            ]);
+
+            a[i] = x[i].wrapping_add(h[i]);
+            b[i] = y[i].wrapping_add(k[i]);
+        }
+
+        for i in 0..4 {
+            p[i] = a[i]
+                .wrapping_add(a[(i + 1) % 4])
+                .wrapping_add(a[(i + 2) % 4]);
+            q[i] = b[(i + 1) % 4]
+                .wrapping_add(b[(i + 2) % 4])
+                .wrapping_add(b[(i + 3) % 4]);
+        }
+
+        let block_res = [
+            a[0] as u64 * b[0] as u64,
+            a[1] as u64 * b[1] as u64,
+            a[2] as u64 * b[2] as u64,
+            a[3] as u64 * b[3] as u64,
+            p[0] as u64 * q[0] as u64,
+            p[1] as u64 * q[1] as u64,
+            p[2] as u64 * q[2] as u64,
+            p[3] as u64 * q[3] as u64,
+        ];
+
+        self.block_sums
+            .iter_mut()
+            .zip(block_res.iter())
+            .for_each(|(sum, &res)| {
+                *sum = sum.wrapping_add(res);
+            });
+
+        self.block_index += 1;
+    }
+
+    fn finalize(&self, out: &mut digest::Output<Self>) {
+        for (i, block) in self.block_sums.iter().enumerate() {
+            let bytes = block.to_ne_bytes(); // Convert u64 to little-endian byte array
+            for (j, &byte) in bytes.iter().enumerate() {
+                out[i * 8 + j] = byte;
+            }
+        }
+    }
+}
+
+impl KeySizeUser for MultimixerCore {
+    type KeySize = U32;
+
+    fn key_size() -> usize {
+        Self::KeySize::USIZE
+    }
+}
+
+impl KeyInit for MultimixerCore {
+    //Uses the key to initialize ChaCha8Rng RNG and fills the key_blocks array
+    fn new(key: &Key<Self>) -> Self {
+        Self {
+            block_sums: [0; 8],
+            key_blocks: Vec::new(),
+            block_index: 0,
+            rng: Some(ChaCha8Rng::from_seed(
+                key.as_slice()
+                    .try_into()
+                    .expect("Key needs to be able to use as seed."),
+            )),
+        }
+    }
+
+    //Uses key instead of RNG, needs to be same size as message.
+    fn new_from_slice(key: &[u8]) -> Result<Self, InvalidLength> {
```

Comment:
In general, `T::new(&key)` and `T::new_from_slice(&key).unwrap()` should produce the same result for the same `key`.